### PR TITLE
Vamshi/edp bug

### PR DESCRIPTION
**Pull request by fazerlicourice7**
nx.dag_longest_path and explicit enumeration of all paths gives different results.
Now using explicit enumeration to line up with symbolic simulate. Almost perfect alignment now. Remnants likely floating point error.

### CODE REVIEW

File: src/sim_util.py [hunk 1]
```diff
@@ -183,9 +183,12 @@ def update_schedule_with_latency(schedule, latency):
     for edge in schedule.edges:
         node = edge[0]
         func = schedule.nodes.data()[node]["function"]
+        scaling = 1
         if func == "stall":
             func = node.split("_")[3]
-        schedule.edges[edge]["weight"] = latency[func]
+        if func == "Buf" or func == "MainMem":
+            scaling = 16
```
Comment (patrickmcewen) [inline src/sim_util.py:190]: why 16 here? does this have anything to do with hardcoded value on 411?

**Comment by patrickmcewen** [comment]
great to hear that we are getting closer on alignment!

**Review by patrickmcewen — APPROVED** [review]

**Comment by fazerlicourice7** [comment]
Sample output:
```
Initial EDP: 40218295.06237531 E-18 Js. Active Energy: 0.05142865898112112 nJ. Passive Energy: 50226.724287235695 nJ. Execution time: 800.7341599999997 ns

Running Inverse Pass
Initial EDP: 40177591.0810017 E-18 Js. Active Energy: 0.0514286589811200 nJ. Passive Energy: 50175.8909602628 nJ. Execution time: 800.734160000000 ns
Final EDP  : 25110994.4258692 E-18 Js. Active Energy: 0.0194273110741270 nJ. Passive Energy: 36212.5398093419 nJ. Execution time: 693.433299253061 ns

Running Forward Pass
longest path after update latency: 754.9588772514956
Initial EDP: 25141520.402286083 E-18 Js. Active Energy: 0.019427311074126936 nJ. Passive Energy: 36256.56131285172 nJ. Execution time: 693.4332992530611 ns
Final EDP  : 25141520.402286083 E-18 Js. Active Energy: 0.019427311074126936 nJ. Passive Energy: 36256.56131285172 nJ. Execution time: 693.4332992530611 ns

Running Inverse Pass
Initial EDP: 25110994.4258692 E-18 Js. Active Energy: 0.0194273110741270 nJ. Passive Energy: 36212.5398093419 nJ. Execution time: 693.433299253061 ns
Final EDP  : 15694371.5162449 E-18 Js. Active Energy: 0.00825479413507448 nJ. Passive Energy: 26067.7706394990 nJ. Execution time: 602.060174742419 ns

Running Forward Pass
longest path after update latency: 656.8506731073994
Initial EDP: 15717382.753965415 E-18 Js. Active Energy: 0.008254794135074494 nJ. Passive Energy: 26105.991466396215 nJ. Execution time: 602.0601747424195 ns
Final EDP  : 15717382.753965415 E-18 Js. Active Energy: 0.008254794135074494 nJ. Passive Energy: 26105.991466396215 nJ. Execution time: 602.0601747424195 ns

Running Inverse Pass
Initial EDP: 15694371.5162449 E-18 Js. Active Energy: 0.00825479413507448 nJ. Passive Energy: 26067.7706394990 nJ. Execution time: 602.060174742419 ns
Final EDP  : 9808982.19766231 E-18 Js. Active Energy: 0.00344332433675511 nJ. Passive Energy: 18678.2162763863 nJ. Execution time: 525.156162892287 ns

Running Forward Pass
longest path after update latency: 574.0658414322791
Initial EDP: 9826490.220774403 E-18 Js. Active Energy: 0.003443324336755117 nJ. Passive Energy: 18711.554975137708 nJ. Execution time: 525.1561628922873 ns
Final EDP  : 9826490.220774403 E-18 Js. Active Energy: 0.003443324336755117 nJ. Passive Energy: 18711.554975137708 nJ. Execution time: 525.1561628922873 ns

Running Inverse Pass
Initial EDP: 9808982.19766231 E-18 Js. Active Energy: 0.00344332433675511 nJ. Passive Energy: 18678.2162763863 nJ. Execution time: 525.156162892287 ns
Final EDP  : 6130613.87355113 E-18 Js. Active Energy: 0.00165844614260453 nJ. Passive Energy: 13352.6478939692 nJ. Execution time: 459.130890051866 ns

Running Forward Pass
longest path after update latency: 502.4809476226584
Initial EDP: 6143996.248813322 E-18 Js. Active Energy: 0.0016584461426045276 nJ. Passive Energy: 13381.79508391476 nJ. Execution time: 459.13089005186595 ns
Final EDP  : 6143996.248813322 E-18 Js. Active Energy: 0.0016584461426045276 nJ. Passive Energy: 13381.79508391476 nJ. Execution time: 459.13089005186595 ns

Running Inverse Pass
Initial EDP: 6130613.87355113 E-18 Js. Active Energy: 0.00165844614260453 nJ. Passive Energy: 13352.6478939692 nJ. Execution time: 459.130890051866 ns
Final EDP  : 3831633.67088236 E-18 Js. Active Energy: 0.000725420380596723 nJ. Passive Energy: 9514.33998943261 nJ. Execution time: 402.721931631136 ns

Running Forward Pass
longest path after update latency: 441.1749103659024
Initial EDP: 3841929.723204882 E-18 Js. Active Energy: 0.0007254203805967247 nJ. Passive Energy: 9539.906146907126 nJ. Execution time: 402.72193163113593 ns
Final EDP  : 3841929.723204882 E-18 Js. Active Energy: 0.0007254203805967247 nJ. Passive Energy: 9539.906146907126 nJ. Execution time: 402.72193163113593 ns
```

Still some discrepancy in passive energy

**Comment by fazerlicourice7** [comment]
Need to debug more why `dag_longest_path` doesn't work. but going to merge in the explicit one for now.

File: src/sim_util.py (hunk 1 above)
Comment (fazerlicourice7) [reply src/sim_util.py]: related. But good catch. This I could actually update back to `data['size']`.
On line 411, the node data doesn't yet have the size attribute, so I had to hardcode the 16 for all values. Need to at some point go back and fix that and have the data sizes loaded from the instrumentation, but for now they're all set as 16 bits.

**Comment by patrickmcewen** [comment]
> Sample output:
> 
> ```
> Initial EDP: 40218295.06237531 E-18 Js. Active Energy: 0.05142865898112112 nJ. Passive Energy: 50226.724287235695 nJ. Execution time: 800.7341599999997 ns
> 
> Running Inverse Pass
> Initial EDP: 40177591.0810017 E-18 Js. Active Energy: 0.0514286589811200 nJ. Passive Energy: 50175.8909602628 nJ. Execution time: 800.734160000000 ns
> Final EDP  : 25110994.4258692 E-18 Js. Active Energy: 0.0194273110741270 nJ. Passive Energy: 36212.5398093419 nJ. Execution time: 693.433299253061 ns
> 
> Running Forward Pass
> longest path after update latency: 754.9588772514956
> Initial EDP: 25141520.402286083 E-18 Js. Active Energy: 0.019427311074126936 nJ. Passive Energy: 36256.56131285172 nJ. Execution time: 693.4332992530611 ns
> Final EDP  : 25141520.402286083 E-18 Js. Active Energy: 0.019427311074126936 nJ. Passive Energy: 36256.56131285172 nJ. Execution time: 693.4332992530611 ns
> 
> Running Inverse Pass
> Initial EDP: 25110994.4258692 E-18 Js. Active Energy: 0.0194273110741270 nJ. Passive Energy: 36212.5398093419 nJ. Execution time: 693.433299253061 ns
> Final EDP  : 15694371.5162449 E-18 Js. Active Energy: 0.00825479413507448 nJ. Passive Energy: 26067.7706394990 nJ. Execution time: 602.060174742419 ns
> 
> Running Forward Pass
> longest path after update latency: 656.8506731073994
> Initial EDP: 15717382.753965415 E-18 Js. Active Energy: 0.008254794135074494 nJ. Passive Energy: 26105.991466396215 nJ. Execution time: 602.0601747424195 ns
> Final EDP  : 15717382.753965415 E-18 Js. Active Energy: 0.008254794135074494 nJ. Passive Energy: 26105.991466396215 nJ. Execution time: 602.0601747424195 ns
> 
> Running Inverse Pass
> Initial EDP: 15694371.5162449 E-18 Js. Active Energy: 0.00825479413507448 nJ. Passive Energy: 26067.7706394990 nJ. Execution time: 602.060174742419 ns
> Final EDP  : 9808982.19766231 E-18 Js. Active Energy: 0.00344332433675511 nJ. Passive Energy: 18678.2162763863 nJ. Execution time: 525.156162892287 ns
> 
> Running Forward Pass
> longest path after update latency: 574.0658414322791
> Initial EDP: 9826490.220774403 E-18 Js. Active Energy: 0.003443324336755117 nJ. Passive Energy: 18711.554975137708 nJ. Execution time: 525.1561628922873 ns
> Final EDP  : 9826490.220774403 E-18 Js. Active Energy: 0.003443324336755117 nJ. Passive Energy: 18711.554975137708 nJ. Execution time: 525.1561628922873 ns
> 
> Running Inverse Pass
> Initial EDP: 9808982.19766231 E-18 Js. Active Energy: 0.00344332433675511 nJ. Passive Energy: 18678.2162763863 nJ. Execution time: 525.156162892287 ns
> Final EDP  : 6130613.87355113 E-18 Js. Active Energy: 0.00165844614260453 nJ. Passive Energy: 13352.6478939692 nJ. Execution time: 459.130890051866 ns
> 
> Running Forward Pass
> longest path after update latency: 502.4809476226584
> Initial EDP: 6143996.248813322 E-18 Js. Active Energy: 0.0016584461426045276 nJ. Passive Energy: 13381.79508391476 nJ. Execution time: 459.13089005186595 ns
> Final EDP  : 6143996.248813322 E-18 Js. Active Energy: 0.0016584461426045276 nJ. Passive Energy: 13381.79508391476 nJ. Execution time: 459.13089005186595 ns
> 
> Running Inverse Pass
> Initial EDP: 6130613.87355113 E-18 Js. Active Energy: 0.00165844614260453 nJ. Passive Energy: 13352.6478939692 nJ. Execution time: 459.130890051866 ns
> Final EDP  : 3831633.67088236 E-18 Js. Active Energy: 0.000725420380596723 nJ. Passive Energy: 9514.33998943261 nJ. Execution time: 402.721931631136 ns
> 
> Running Forward Pass
> longest path after update latency: 441.1749103659024
> Initial EDP: 3841929.723204882 E-18 Js. Active Energy: 0.0007254203805967247 nJ. Passive Energy: 9539.906146907126 nJ. Execution time: 402.72193163113593 ns
> Final EDP  : 3841929.723204882 E-18 Js. Active Energy: 0.0007254203805967247 nJ. Passive Energy: 9539.906146907126 nJ. Execution time: 402.72193163113593 ns
> ```
> 
> Still some discrepancy in passive energy

Interesting that latency lines up exactly. In the past when I've seen passive energy discrepancy it was usually because of latency discrepancy but not the case here.

**Pull request merged**